src/ext/php7/engine_hooks.c

```diff
@@ -479,10 +448,13 @@ static int _dd_opcode_default_dispatch(zend_execute_data *execute_data) {
     return ZEND_USER_OPCODE_DISPATCH;
 }
 
-int ddtrace_wrap_fcall(zend_execute_data *execute_data) {
-    zend_function *current_fbc = NULL;
+static int _dd_begin_fcall_handler(zend_execute_data *execute_data) {
+    zend_function *current_fbc = EX(call)->func;
+    if (!current_fbc) {
+        return _dd_opcode_default_dispatch(execute_data);
+    }
     ddtrace_dispatch_t *dispatch = NULL;
-    if (!_dd_should_trace_call(execute_data, &current_fbc, &dispatch)) {
+    if (!_dd_should_trace_call(EX(call), current_fbc, &dispatch)) {
         return _dd_opcode_default_dispatch(execute_data);
     }
     int vm_retval = _dd_opcode_default_dispatch(execute_data);
```

Comment:
Let's see if we can rework this bit so that `_dd_opcode_default_dispatch` only shows up in this section of code once.

Reply:
For unaltered dispatch, we should call our neighbor _at the end_. Since we don't mutate any VM state, we are free to do our own thing and then call them -- this way they have less chance of screwing up what we want to do.

For legacy dispatch, we need to call the neighbor first or not at all.